src/merkle_tree.rs

```diff
@@ -7,34 +9,202 @@ pub struct MerkleTree {
     layers: Vec<Vec<Hash>>,
 }
 
+impl Display for MerkleTree {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for (idx, layer) in self.layers.iter().enumerate() {
+            writeln!(f, "Layer: {idx}: {{")?;
+            for hash in layer {
+                writeln!(f, "\t{}", hex::encode(hash))?;
+            }
+            writeln!(f, "}}")?;
+        }
+        Ok(())
+    }
+}
+
 impl MerkleTree {
     pub fn from_leaves<T: std::convert::AsRef<[u8]>>(leaves: &[T]) -> Self {
         let mut tree = Self::default();
-        tree.build_first_layer(leaves);
+        tree.layers.push(Self::build_first_layer(leaves));
+        while let Some(previous_layer) = tree.layers.first() {
+            if previous_layer.len() <= 1 {
+                break;
+            }
```

Comment:
Can we do `while let Some(previous_layer) = tree.layers.first() && previous_layer.len() <= 1` here? Not sure if it works